[GlobalDCE] Support ptrauth in VFE

kubamracek opened the pull request:
Will add a testcase, but in summary, ptrauth function pointers in vtables need extra support in Virtual Function Elimination in GlobalDCE. The vtable in this case contains a reference not to the target function, but to the .ptrauth global variable (in llvm.ptrauth section), so we need to allow non-function references to participate in VFE. Consequence of that is that after VFE removes globals, we need to account for dead globals that have remaining references from vtables (just like we account for dead funcs that have remaining references from vtables).

rdar://85970023

kubamracek commented:
@swift-ci please test

fhahn commented:
This cannot be submitted directly to open-source because the ptrauth changes are on this fork only for now, right?

kubamracek commented:
Correct. Most of LLVM ptrauth support is downstream (on this fork).

kubamracek committed:
[GlobalDCE] Support ptrauth in VFE

Ptrauth function pointers in vtables need extra support in Virtual Function Elimination in GlobalDCE. The vtable in this case contains a reference not to the target function, but to the .ptrauth global variable (in llvm.ptrauth section), so we need to allow non-function references to participate in VFE. Consequence of that is that after VFE removes globals, we need to account for dead globals that have remaining references from vtables (just like we account for dead funcs that have remaining references from vtables).
